File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlD4MissingRequestHandlerScenario.cs
// Verifies RL-D4 Missing Request Handler behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-D4 verifies public failure and server evidence for a missing request handler.
internal static class RlD4MissingRequestHandlerScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var failed = await consumer.Post("/profile/request/missing")
            .Body(new ProfileReq("fast", "rl-d4-missing"))
            .AsyncRaw();
        ZlinkStreamAssert.Ensure(failed.Status >= 500, "RL-D4 expected public failure for missing request handler.");

        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var request = new EvidenceWaitReq(["dispatch-error|", "packet=MissingProfileReq"], []);
            var waitA = providerA.Post("/evidence/wait").Body(request).Async<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(request).Async<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ZlinkStreamAssert.Ensure(
                evidence.Any(line => line.Contains("dispatch-error|", StringComparison.Ordinal)
                                     && line.Contains("packet=MissingProfileReq", StringComparison.Ordinal)),
                "RL-D4 dispatch-error marker missing.");
        }

        Console.WriteLine("scenario RL-D4 passed");
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-d4-missing-request-handler-scenario.ts
// RL-D4: Same-version peers가 public error kind를 보존한다 시나리오를 검증한다.
import type { RequestFailureRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { waitForProviderEvidenceLine } from '../Support/provider-evidence';
import { profileReq, waitForAnyProviderTraffic } from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';

export async function runRlD4(options: ClientOptions): Promise<void> {
  await waitForAnyProviderTraffic(options.consumerUrl, 'rl-d4-ready');
  const failed = await postJson<RequestFailureRes>(
    options.consumerUrl,
    '/profile/missing-request',
    profileReq('rl-d4-missing')
  );
  ensure(failed.failed, 'RL-D4 expected public failure for missing request handler.');
  ensure(
    failed.failureMessage.includes('request handler is registered'),
    'RL-D4 decoded errorMessage did not preserve the missing-handler message.'
  );

  const line = await waitForProviderEvidenceLine(
    options,
    (entry) => entry.includes('dispatch-error|') && entry.includes('packet_name=MissingProfileReq'),
    'RL-D4 dispatch-error evidence timed out.'
  );
  ensure(
    line.includes('dispatch-error|')
      && line.includes('outcome=failed')
      && line.includes('reason=no_handler')
      && line.includes('action=reply_error')
      && line.includes('packet_name=MissingProfileReq'),
    'RL-D4 dispatch-error marker missing.'
  );

  const followUp = await postJson<{ readonly value: string }>(
    options.consumerUrl,
    '/profile/request',
    profileReq('rl-d4-follow-up')
  );
  ensure(followUp.value === 'profile:fast', 'RL-D4 success Response follow-up failed.');

  console.log('scenario RL-D4 passed');
}
